src/main/java/duke/parser/Parser.java

```diff
@@ -0,0 +1,154 @@
+        } else if (input.startsWith("find")) {
+            String target = parseFind(input);
+            return new FindCommand(target);
+        } else {
+            if (!input.startsWith("todo") && !input.startsWith("deadline") && !input.startsWith("event")) {
+                throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+            }
+            Task task;
+            if (input.startsWith("todo")) {
+                task = parseTodo(input);
+            }
+            else if (input.startsWith("deadline")) {
+                task = parseDeadline(input);
+            }
+            else {
+                task = parseEvent(input);
+            }
+            return new AddCommand(task);
+        }
+    }
+
+    /**
+     * Parses users inputs that wishes to mark a task as done
+     *
+     * @param input the user's input
+     * @return the index of the task user wishes to mark as done
+     * @throws DukeException if the user's input is in an wrong format
+     */
+    public static int parseDone(String input) throws DukeException {
+        String numberString = String.valueOf(input.charAt(5));
+        if (!pattern.matcher(numberString).matches()) {
+            throw new DukeException("Please enter \'done [index of item]\' to mark item as done.");
+        }
+        int number = Integer.parseInt(String.valueOf(numberString));
+        if (number > tasks.getSize() || number < 0) {
+            throw new DukeException("Item does not exist, we cannot mark it as done.");
+        }
+        Task currTask = tasks.get(number - 1);
+        if (currTask.getStatusIcon().equals(String.valueOf('X'))) {
+            throw new DukeException("Item is already marked as done, we cannot mark it as done again.");
+        }
+        return number - 1;
+    }
+
+    /**
+     * Parses users inputs that wishes to delete a task
+     *
+     * @param input the user's input
+     * @return the index of the task user wishes to delete
+     * @throws DukeException if the user's input is in an wrong format
+     */
+    public static int parseDelete(String input) throws DukeException {
+        String numberString = String.valueOf(input.charAt(7));
+        if (!pattern.matcher(numberString).matches()) {
+            throw new DukeException("Please enter \'delete [index of item]\' to mark item as done.");
+        }
+        int number = Integer.parseInt(String.valueOf(numberString));
+        if (number > tasks.getSize() || number < 0) {
+            throw new DukeException("Item does not exist, we cannot delete it.");
+        }
+        return number - 1;
+    }
+
+    /**
+     * Parses users inputs that wishes to add a todo task
+     *
+     * @param input the user's input
+     * @return the task that the user added
+     * @throws DukeException if the user's input is in an wrong format
+     */
+    public static String parseFind(String input) throws DukeException {
+        if (input.length() == 4) {
```

Comment:
I think it's advisable not to check for validity using String length in this case, especially if you're not sanitising the input using .trim() . For example, if the input was "todo " with whitespaces at the end, it should be considered an invalid input, but I believe your code does not catch it.